nixos/modules/system/boot/loader/limine/limine-install.py

```diff
@@ -122,18 +131,72 @@ class BootSpec:
     label: str
     toplevel: str
     specialisations: Dict[str, "BootSpec"]
+    xen: XenBootSpec
     initrd: str | None = None
     initrdSecrets: str | None = None
 
-
 def bootjson_to_bootspec(bootjson: dict) -> BootSpec:
     specialisations = bootjson['org.nixos.specialisation.v1']
     specialisations = {k: bootjson_to_bootspec(v) for k, v in specialisations.items()}
+    xen = bootjson['org.xenproject.bootspec.v1']
     return BootSpec(
         **bootjson['org.nixos.bootspec.v1'],
         specialisations=specialisations,
+        xen=xen,
     )
 
+def xen_config_entry(
+    levels: int, bootspec: BootSpec, xenVersion: str, gen: str, time: str
+) -> str:
+    """Generate EFI and BIOS entries for Xen dom0 kernels.
+
+    Arguments:
+    levels -- The number of Limine menu levels for entries
+    bootspec -- The NixOS BootSpec used for generating this Limine configuration
+    xenVersion -- The version of Xen the entry is generated for, from the boot extension
+    gen -- The system generation these entries are generated for
+    time -- The build time for the configuration
+    """
+    entry = ""
+    # generate entry common header for both BIOS and EFI entries
+    entry_header = "/" * levels + label + "\n"
+    entry_header += f"comment: {bootspec.label}, built on {time}\n"
+    # generate entry common footer for both BIOS and EFI entries
+    entry_footer = ""
+    # set xenParams as the multiboot executable's parameters
+    if "xenParams" in bootspec.xen and len(bootspec.xen["xenParams"]) > 0:
+        # TODO: Understand why the first argument is ignored below?
+        # --- to work around first argument being ignored
+        entry_footer += (
+            "cmdline: -- " + " ".join(bootspec.xen["xenParams"]).strip() + "\n"
+        )
+    # load the linux kernel as the second module
+    entry_footer += "module_path: " + get_kernel_uri(bootspec.kernel) + "\n"
+    # set kernel parameters as the parameters to the first module
+    # TODO: Understand why the first argument is ignored below?
+    # --- to work around first argument being ignored
+    entry_footer += (
+        "module_string: -- "
+        + " ".join(["init=" + bootspec.init] + bootspec.kernelParams).strip()
+        + "\n"
+    )
+    if bootspec.initrd:
+        # the final module is the initrd
+        entry_footer += "module_path: " + get_kernel_uri(bootspec.initrd) + "\n"
+    # load Xen dom0 as the executable, generate entries for EFI and BIOS boot
+    if config("efiSupport") and "xenEfi" in boot_spec.xen:
+        entry += entry_header
+        entry += "protocol: multiboot2\n"
+        entry += "path: " + get_copied_path_uri(bootspec.xen["xenEfi"], "xen") + "\n"
+        entry += entry_footer
```

Comment:
xenEfi is supposed to be a EFI executable, eh? Why are we using multiboot2 to boot it? Maybe try `protocol: efi`? I am currently rebuilding my system to test if multiboot2 works then I'll try efi chainloading if it doesn't (which is what I expect)

Reply:
Unfortunately I don't have hardware free for this, but yes it's possible it should be the multiboot binary, or the EFI UKI directly as an EFI payload, I'm not sure what will work best. I elaborated below on the usage of multiboot1 however, as multiboot2 did not work under Limine for the Xen multiboot binary when BIOS booting.

Reply:
After some rummaging this morning I've turned up an ASRock board with a Ryzen CPU which has stock UEFI firmware. I'll test it out and update the PR based on what works. If the multiboot executable works as-is I'll be able to simplify this code significantly, if it needs the EFI executable or protocol: EFI then I'll update the script to accommodate that.